system-parachains/encointer/src/migrations_fix.rs

```diff
@@ -17,173 +17,131 @@
-			}
-
-			#[cfg(feature = "try-runtime")]
-			fn post_upgrade(_state: sp_std::vec::Vec<u8>) -> Result<(), TryRuntimeError> {
-				ensure!(StorageVersion::get::<Pallet<T>>() == 1, "Must upgrade");
-				Ok(())
-			}
-		}
-	}
-}
-pub mod scheduler {
-	// this is necessary because migrations from v0 to v3 are no longer available in the scheduler
-	// pallet code and migrating is only possible from v3. The strategy here is to empty the agenda
-	// (has been empty since genesis)
-	use frame_support::traits::OnRuntimeUpgrade;
-	use frame_system::pallet_prelude::BlockNumberFor;
-	use pallet_scheduler::*;
-	use sp_std::vec::Vec;
-
-	#[cfg(feature = "try-runtime")]
-	use sp_runtime::TryRuntimeError;
-
-	/// The log target.
-	const TARGET: &'static str = "runtime::fix::scheduler::migration";
-
-	pub mod v1 {
-		use super::*;
-		use frame_support::{pallet_prelude::*, traits::schedule};
-
-		#[cfg_attr(any(feature = "std", test), derive(PartialEq, Eq))]
-		#[derive(Clone, RuntimeDebug, Encode, Decode)]
-		pub(crate) struct ScheduledV1<Call, BlockNumber> {
-			maybe_id: Option<Vec<u8>>,
-			priority: schedule::Priority,
-			call: Call,
-			maybe_periodic: Option<schedule::Period<BlockNumber>>,
-		}
-
-		#[frame_support::storage_alias]
-		pub(crate) type Agenda<T: Config> = StorageMap<
-			Pallet<T>,
-			Twox64Concat,
-			BlockNumberFor<T>,
-			Vec<Option<ScheduledV1<<T as Config>::RuntimeCall, BlockNumberFor<T>>>>,
-			ValueQuery,
-		>;
-
-		#[frame_support::storage_alias]
-		pub(crate) type Lookup<T: Config> =
-			StorageMap<Pallet<T>, Twox64Concat, Vec<u8>, TaskAddress<BlockNumberFor<T>>>;
-	}
-
-	pub mod v3 {
-		use super::*;
-		use frame_support::pallet_prelude::*;
-
-		#[frame_support::storage_alias]
-		pub(crate) type Agenda<T: Config> = StorageMap<
-			Pallet<T>,
-			Twox64Concat,
-			BlockNumberFor<T>,
-			Vec<Option<ScheduledV3Of<T>>>,
-			ValueQuery,
-		>;
-
-		#[frame_support::storage_alias]
-		pub(crate) type Lookup<T: Config> =
-			StorageMap<Pallet<T>, Twox64Concat, Vec<u8>, TaskAddress<BlockNumberFor<T>>>;
-	}
-
-	pub mod v4 {
-		use super::*;
-		use frame_support::pallet_prelude::*;
-
-		#[frame_support::storage_alias]
-		pub type Agenda<T: Config> = StorageMap<
-			Pallet<T>,
-			Twox64Concat,
-			BlockNumberFor<T>,
-			BoundedVec<
-				Option<ScheduledOf<T>>,
-				<T as pallet_scheduler::Config>::MaxScheduledPerBlock,
-			>,
-			ValueQuery,
-		>;
-
-		#[cfg(feature = "try-runtime")]
-		pub(crate) type TaskName = [u8; 32];
 
-		#[cfg(feature = "try-runtime")]
-		#[frame_support::storage_alias]
-		pub(crate) type Lookup<T: Config> =
-			StorageMap<Pallet<T>, Twox64Concat, TaskName, TaskAddress<BlockNumberFor<T>>>;
+				let mut invulnerables: Vec<<T as frame_system::Config>::AccountId> =
+					raw_keys.iter().map(|&pk| pk.into()).collect();
+				invulnerables.sort();
+				let invulnerables: BoundedVec<_, T::MaxInvulnerables> =
+					invulnerables.try_into().unwrap();
+				pallet_collator_selection::Invulnerables::<T>::put(invulnerables);
```

Comment:
This is not technically wrong but I still think it is bad practice for the on-chain `AccountId` to be the same key as the Aura key.

The on-chain account should be a long-standing key held in a wallet with backups. They can only be changed via governance motion.

The session key is a "hot" in-node key used for signing blocks. It is not meant to be long-standing. For example, if one of the invulnerables wants to start a new collator machine, standard practice would be to start the second machine, call `rotate_keys`, and set the new session keys to correspond to the unchanged on-chain account. The old ones would disappear when killing the old machine.

This pattern prevents this, either the keys in the migration are session keys and potentially not properly backed up or they are wallet keys that collators will need to inject into online collator nodes to get started before rotating.

Reply:
It depends what you mean by "collator keys" here.

- Whatever values you `put` into `pallet_collator_selection::Invulnerables::<T>::put(invulnerables);` should be keys that are in a wallet/cold storage, as in keys you want to live for a long time and would expect to hold funds and make transactions.
- `Validators` should be a vec of the account keys (wallets).
- `QueuedKeys` and `NextKeys` should be the map of `ValidatorId => SessionKey`.
- `KeyOwner` should be the map of `(AURA, SessionKey) => ValidatorId`.

Maybe @georgepisaltu can double check this from pallet-session?

Reply:
this sounds coherent. Did it like this in my most recent commits